### PR TITLE
Feature/x ref support

### PR DESCRIPTION
**Pull request by mcnallys**

### CODE REVIEW

File: src/PDFsharper.UnitTests/Pdf.AcroForms/PdfAcroFieldTests.cs [hunk 1]
```diff
@@ -78,43 +78,7 @@ public void GetIsVisible_WithFlagsRemoved()
             Assert.IsFalse(isNotVisible, "isVisible should be false");
         }
 
-        [TestMethod]
-        public void RemoveJavascript()
```
Comment (pex-simsr) [inline src/PDFsharper.UnitTests/Pdf.AcroForms/PdfAcroFieldTests.cs:83]: Why did these test methods get pulled?

File: src/PDFsharper.UnitTests/Pdf.AcroForms/PdfAcroFieldTests.cs [hunk 2]
```diff
@@ -78,43 +78,7 @@ public void GetIsVisible_WithFlagsRemoved()
...
-            Assert.IsTrue(document.Catalog.Elements.GetDictionary("/Names").Elements.Any(e => e.Key == "/JavaScript"), "document-catalog should have a Javascript item");
-
-            field.RemoveJavascript();
 
-            Assert.IsTrue(document.Catalog.Elements.Any(e => e.Key == "/Names"), "document should have a Names item");
-            Assert.IsFalse(document.Catalog.Elements.GetDictionary("/Names").Elements.Any(e => e.Key == "/JavaScript"), "document-catalog should not have a Javascript item");
-        }
-
-        [TestMethod]
-        public void RemoveJavascript_NoJavascript()
```
Comment (pex-simsr) [inline src/PDFsharper.UnitTests/Pdf.AcroForms/PdfAcroFieldTests.cs:108]: same..?

File: src/PDFsharper.UnitTests/Pdf.IO/ParserTests.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+namespace PDFsharper.UnitTests.Pdf.IO
+{
+    [TestClass]
+    public class ParserTests
+    {
+
+#if DEBUG
+        [TestMethod]
+        public void Debug_FileRead()
```
Comment (pex-simsr) [inline src/PDFsharper.UnitTests/Pdf.IO/ParserTests.cs]: Anyone who pulls and doesn't have these documents will have an issue running these tests.

File: src/PdfSharper/Pdf.Advanced/PdfCrossReferenceStream.cs [hunk 4]
```diff
@@ -49,11 +53,39 @@ public PdfCrossReferenceStream(PdfDocument document)
...
+                Stream = new PdfStream(this);
+            }
+            finally
+            {
+                Owner.UnderConstruction = false;
+            }
         }
 
-        public readonly List<CrossReferenceStreamEntry> Entries = new List<CrossReferenceStreamEntry>();
+        public List<CrossReferenceStreamEntry> Entries { get; private set; } = new List<CrossReferenceStreamEntry>();
```
Comment (pex-simsr) [inline src/PdfSharper/Pdf.Advanced/PdfCrossReferenceStream.cs]: I learn something everyday.. this must be new new feature that we talked about the other day.

**Review by pex-simsr — DISMISSED** [review]

File: src/PDFsharper.UnitTests/Pdf.AcroForms/PdfAcroFieldTests.cs (hunk 2 above)
Comment (mcnallys) [reply src/PDFsharper.UnitTests/Pdf.AcroForms/PdfAcroFieldTests.cs:108]: I removed the method. Still working this PR btw.

File: src/PDFsharper.UnitTests/Pdf.AcroForms/PdfAcroFieldTests.cs (hunk 1 above)
Comment (mcnallys) [reply src/PDFsharper.UnitTests/Pdf.AcroForms/PdfAcroFieldTests.cs:83]: The javascript based error only happens when all fields are removed from the form. These methods were pulled for that reason and for performance, flattening with these in place was sluggish.

File: src/PDFsharper.UnitTests/Pdf.IO/ParserTests.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+using PdfSharper.Pdf.Internal;
+using System.Diagnostics;
+
+namespace PDFsharper.UnitTests.Pdf.IO
+{
+    [TestClass]
+    public class ParserTests
+    {
+
+#if DEBUG && PEX
```
Comment (mcnallys) [inline src/PDFsharper.UnitTests/Pdf.IO/ParserTests.cs:19]: These are only for running here - i can remove them from the solution entirely if deemed appropriate.

File: src/PdfSharper/Pdf/PdfDocument.cs [hunk 6]
```diff
@@ -799,6 +799,9 @@ private PdfReference[] GatherFirstPageAndDocumentReferences()
 
             var firstPageObjects = PdfTraversalUtility.TransitiveClosure(Pages[0], exclusions).Select(kvp => kvp.Key).ToList();
 
+            //sometimes the transitiveclosure does not include the root element
+            firstPageObjects.Add(Pages[0].Reference);
```
Comment (pex-simsr) [inline src/PdfSharper/Pdf/PdfDocument.cs]: because of the "sometimes", should we make sure it isn't there before adding - prevent duplicate entries?

**Review by pex-simsr — DISMISSED** [review]

**Review by pex-goldenc — DISMISSED** [review]
Coding looks fine. Tests I've built pass with known-good documents. Can't swear it's perfect, but it looks good.

**Review by pex-simsr — DISMISSED** [review]

**Review by pex-goldenc — APPROVED** [review]
Updates are sensible.

**Pull request merged**